### PR TITLE
core, core/state, core/vm: removal of exported account getters

### PR DESCRIPTION
**Pull request by obscuren**
Removed exported statedb object accessors, reducing the chance for nasty
bugs to creep in. It's also ugly and unnecessary to have these methods.

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -209,7 +209,8 @@ func (self *BlockChain) loadLastState() error {
 		return err
 	}
 	self.stateCache = statedb
-	self.stateCache.GetAccount(common.Address{})
+	// XXX Reviewers: what's this for? I need to know if I need to backport this somehow
```
Comment (Arachnid) [inline core/blockchain.go]: My guess would be that some other code depends (or depended) on that address being in the state cache for some reason.

File: core/vm/contract.go [hunk 2]
```diff
@@ -25,11 +25,13 @@ import (
...
 type ContractRef interface {
 	Address() common.Address
-	Value() *big.Int
-	SetCode(common.Hash, []byte)
-	ForEachStorage(callback func(key, value common.Hash) bool)
 }
 
+type AccountRef common.Address
+
+func (ar AccountRef) Address() common.Address  { return (common.Address)(ar) }
```
Comment (Arachnid) [inline core/vm/contract.go]: Why use this instead of just `Address`?

File: core/vm/contract.go [hunk 3]
```diff
@@ -25,11 +25,13 @@ import (
...
+func Reference(addr common.Address) AccountRef { return (AccountRef)(addr) }
```
Comment (Arachnid) [inline core/vm/contract.go]: The parens here are a bit misleading; it's a constructor, not a C-style cast. Also, couldn't you just do it without the function?

File: core/vm/logger.go [hunk 4]
```diff
@@ -144,11 +144,14 @@ func (l *StructLogger) CaptureState(env *EVM, pc uint64, op OpCode, gas, cost *b
 			storage = make(Storage)
 			// Get the contract account and loop over each storage entry. This may involve looping over
 			// the trie and is a very expensive process.
+
```
Comment (Arachnid) [inline core/vm/logger.go]: This needs fixing before submission, surely?

Comment (obscuren) [inline core/vm/logger.go]: :flushed:

File: core/blockchain.go (hunk 1 above)
Comment (obscuren) [inline core/blockchain.go]: @fjl @karalabe do you know maybe?

File: core/vm/contract.go (hunk 2 above)
Comment (obscuren) [inline core/vm/contract.go]: I'll fix up the docs. This implements an interface.

File: core/blockchain.go (hunk 1 above)
Comment (fjl) [inline core/blockchain.go]: This is not needed anymore.

Comment (fjl) [inline core/blockchain.go]: It was added by @karalabe to bootstrap a previous version of the account cache.

File: core/vm/contract.go [hunk 5]
```diff
@@ -25,11 +25,23 @@ import (
...
+// proves difficult because of the cached jump destinations which
+// are fetched from the parent contract (i.e. the caller), which
+// is a ContractRef.
+type AccountRef common.Address
+
+// Address casts AccountRef to a Address
+func (ar AccountRef) Address() common.Address { return (common.Address)(ar) }
+
+// Reference returns a casted Address to AccountRef
+func Reference(addr common.Address) AccountRef { return (AccountRef)(addr) }
```
Comment (fjl) [inline core/vm/contract.go]: This is not necessary. You can use `AccountRef` like a normal function, e.g. `vm.AccountRef(addr)`.

File: core/state/state_object.go [hunk 6]
```diff
@@ -347,42 +347,42 @@ func (self *StateObject) setCode(codeHash common.Hash, code []byte) {
...
 // as a vm.Account interface that also satisfies the vm.ContractRef
 // interface. Interfaces are awesome.
-func (self *StateObject) Value() *big.Int {
-	panic("Value on StateObject should never be called")
+func (self *stateObject) Value() *big.Int {
+	panic("Value on stateObject should never be called")
 }
 
-func (self *StateObject) ForEachStorage(cb func(key, value common.Hash) bool) {
+func (self *stateObject) ForEachStorage(cb func(key, value common.Hash) bool) {
```
Comment (fjl) [inline core/state/state_object.go]: Please remove this method. You've added a copy with the same code to StateDB, making this one unreachable.

**Comment by obscuren** [comment]
@fjl updated. PTAL

**Pull request merged**